smalltalksrc/VMMaker/CogObjectRepresentation.class.st

```diff
@@ -872,6 +872,41 @@ CogObjectRepresentation >> genPrimitiveFloatSubtract [
 	^self genDoubleArithmetic: SubRdRd preOpCheck: nil
 ]
 
+{ #category : #'primitive generators' }
+CogObjectRepresentation >> genPrimitiveFormat [
+
+	| jumpIsImm primitiveFailure |
+	"Only works for the receiver for now"
+	cogit methodNumArgs > 0 ifTrue: [ ^ UnimplementedPrimitive ].
+
+	cogit MoveR: ReceiverResultReg R: TempReg.
+
+	"Immediates don't have format"
+	cogit AndCq: objectMemory tagMask R: ReceiverResultReg.
+	jumpIsImm := cogit JumpNonZero: 0.
+	
+	cogit MoveR: ReceiverResultReg R: TempReg.
+	
+	"We obtain the header of the receiver, and then from the header we obtain the format:"
+	"Get the entire word with the in destReg"
+	cogit MoveMw: 0 r: ReceiverResultReg R: TempReg.
+	"Mask off format"
+	cogit LogicalShiftRightCq: objectMemory formatShift R: TempReg.
+	cogit AndCq: objectMemory formatMask R: TempReg.
+
+	"Now in TempReg we have the 5 bits representing the format of the instance.
+	We have to convert that value to a Pharo Number Object (SmallInteger)"
+	cogit
+		LogicalShiftLeftCq: objectMemory numSmallIntegerTagBits
+		R: TempReg.
+	"To do that, we shift left 3 for the tag and then tag it as 001, which is the tag for SmallInteger"
+	cogit AddCq: 1 R: TempReg.
```

Comment:
I propose to replace the following to make it compatible with 32 bit VMs too.
This also makes the comment redundant.

```suggestion
	self genConvertIntegerToSmallIntegerInReg: TempReg.
```